lib/facter/facts/linux/cloud/provider.rb

```diff
@@ -10,6 +10,9 @@ def call_the_resolver
           provider = case Facter::Util::Facts::Posix::VirtualDetector.platform
                      when 'hyperv'
                        'azure' unless Facter::Resolvers::Az.resolve(:metadata).empty?
+                     when 'kvm'
+                       # hypervisor.kvm fact is what we reuse
+                       Facts::Linux::Hypervisors::Kvm.new.call_the_resolver.value.first[0]
```

Comment:
Is `call_the_resolver.value` or `call_the_resolver.value.first` ever `nil`? It seems like it may be possible if the hypervisor is `virtualbox` or `parallels`:

https://github.com/puppetlabs/facter/blob/636c4955fcd5449d625f8390a99dad6564857e12/lib/facter/facts/linux/hypervisors/kvm.rb#L17

Or if `kvm?` is false, so `fact_value` is `nil` in https://github.com/puppetlabs/facter/blob/636c4955fcd5449d625f8390a99dad6564857e12/lib/facter/facts/linux/hypervisors/kvm.rb#L19-L21

Could we follow the Azure example and rely on the EC2 metadata to detect AWS, something like:

```ruby
when 'kvm'
  'aws' unless Facter::Resolvers::Ec2.resolve(:metadata).empty?
end
```

Also side note, we'll need a similar change for the [Windows cloud provider](https://github.com/puppetlabs/facter/blob/main/lib/facter/facts/windows/cloud/provider.rb).

Reply:
Good call, that looks much more straight forward.